test/lifecycle_test.dart

```diff
@@ -139,7 +140,78 @@ void main() {
+        ..addAll(newProps)
+      );
+
+      const Map expectedState = const {};
+
+      const Map initialContext = const {
+        'foo': false
+      };
+
+      const Map expectedContext = const {
+        'foo': true
+      };
+
+      Map refMap = {
+        'ref': ((ref) => component = ref),
+      };
+
+      // Add the 'ref' prop separately so it isn't an expected prop since React removes it internally
+      var initialPropsWithRef = new Map.from(initialProps)..addAll(refMap);
+      var newPropsWithRef = new Map.from(newPropsWithDefaults)..addAll(refMap);
+
+      // Render the initial instance
+      var mountNode = new DivElement();
+      react_dom.render(ContextWrapper({'foo': false}, LifecycleTestWithContext(initialPropsWithRef)), mountNode);
+
+      // Verify initial context/setup
+      expect(component.lifecycleCalls, equals([
+        matchCall('getInitialState',                      props: initialPropsWithDefaults, context: initialContext),
+        matchCall('componentWillMount',                   props: initialPropsWithDefaults, context: initialContext),
+        matchCall('render',                               props: initialPropsWithDefaults, context: initialContext),
+        matchCall('componentDidMount',                    props: initialPropsWithDefaults, context: initialContext),
+      ]));
+
+      // Clear the lifecycle calls for to not duplicate the initial calls below
+      component.lifecycleCalls.clear();
+
+      // Trigger a re-render with new content
+      react_dom.render(ContextWrapper({'foo': true}, LifecycleTestWithContext(newPropsWithRef)), mountNode);
+
+      // Verify updated context/setup
+      expect(component.lifecycleCalls, equals([
+        matchCall('componentWillReceiveProps',            args: [newPropsWithDefaults],                                 props: initialPropsWithDefaults, context: initialContext),
+        matchCall('componentWillReceivePropsWithContext', args: [newPropsWithDefaults, expectedContext],                props: initialPropsWithDefaults, context: initialContext),
+        matchCall('shouldComponentUpdateWithContext',     args: [newPropsWithDefaults, expectedState, expectedContext], props: initialPropsWithDefaults, context: initialContext),
+        matchCall('componentWillUpdate',                  args: [newPropsWithDefaults, expectedState],                  props: initialPropsWithDefaults, context: initialContext),
+        matchCall('componentWillUpdateWithContext',       args: [newPropsWithDefaults, expectedState, expectedContext], props: initialPropsWithDefaults, context: initialContext),
+        matchCall('render',                                                                                             props: newPropsWithDefaults, context: expectedContext),
+        matchCall('componentDidUpdate',                   args: [initialPropsWithDefaults, expectedState],              props: newPropsWithDefaults, context: expectedContext),
+      ]));
+    });
```

Comment:
Awesome test! Could we add one more that verifies that only the keys present in `childContextKeys` are passed down, though?

Reply:
Setting an extra key to pass but not having the value in the child should be enough. That's incoming.

Reply:
See 282914b.